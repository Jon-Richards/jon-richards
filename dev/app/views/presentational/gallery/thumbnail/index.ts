export { Thumbnail, ThumbnailProps } from './thumbnail';
