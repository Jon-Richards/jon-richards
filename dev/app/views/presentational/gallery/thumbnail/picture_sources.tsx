import * as React from 'react';

/**
 * Renders an array of <source> tags.  There are ways to produce this behavior
 * with a single srcset attribute, but this method offers better readability at
 * runtime.
 */
export function PictureSources(props: PictureSourcesProps): JSX.Element {
  const mapped = props.sources.map(source => (
    <source
      key={source.source}
      media={source.mediaQuery}
      srcSet={source.source}
      data-testid="gallery__thumb__picture-source"
    />
  ));

  const reversed = reverseForMobileFirst(mapped);

  return <React.Fragment>{reversed}</React.Fragment>;
}

export type PictureSourcesProps = {
  sources: PictureSource[];
}

type PictureSource = {
  mediaQuery: string;
  source: string;
}

/**
 * The browser will use whichever <source> tag matches the environment first,
 * so we can't assume that "min-width" will cascade the way it does with CSS.
 * (If the first source represents the smallest width, it will always return
 * true.) This function simply reverses the array so that larger resolutions
 * will be checked first.
 */
function reverseForMobileFirst(sources: JSX.Element[]): JSX.Element[] {
  return sources.reverse();
}
