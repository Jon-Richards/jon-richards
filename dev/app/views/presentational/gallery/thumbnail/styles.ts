import { Breakpoints } from 'Config/styles/breakpoints';
import { Dimensions } from 'Config/styles/dimensions';

/** Styles for the Thumbnails component. */
export const STYLES = {
  root: `
    width: 100%;
    height: auto;
    padding: 0rem;

    @media (min-width: ${Breakpoints[720]}) {
      transition: all 0.2s ease-in-out;
      transform-origin: 50% 50%;
      &:hover, &:active, &:focus, &:focus-within {
        transform: scale(1.1, 1.1);
      }
    }
  `,
  figure: `
    display: flex;
    flex-wrap: wrap;
    width: 100%;
    height: auto;
    padding: 0rem;
    margin: 0rem;
    vertical-align: top;
  `,
  image: `
    display: inline-flex;
    width: 100%;
    height: auto;
  
    @media (min-width: ${Breakpoints[480]}) {
      width: 50%;
    }
  `,
  caption: `
    display: inline-flex;
    width: 100%;
    padding: ${Dimensions.whitespace};
  
    @media (min-width: ${Breakpoints[480]}) {
      width: 50%;
    }
  `
};
