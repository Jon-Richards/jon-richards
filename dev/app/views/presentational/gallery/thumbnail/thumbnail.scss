@import '~scss-standards';

.root {
  display: flex;
  width: 100%;
  height: auto;
  padding: 0rem;
  margin: 0rem 0rem $std-dimension-gap 0rem;

  @media (min-width: std-get-breakpoint(720)) {
    display: inline-flex;
    width: calc(50% - 1rem);
    &:nth-of-type(odd) {
      margin-right: 2rem;
    }
  }
}

.figure {
  display: flex;
  flex-wrap: wrap;
  width: 100%;
  height: auto;
  padding: 0rem;
  margin: 0rem;
  vertical-align: top;
}

.image {
  display: inline-flex;
  width: 100%;
  height: auto;

  @media (min-width: std-get-breakpoint(480)) {
    width: 50%;
  }
}

.caption {
  display: inline-flex;
  width: 100%;
  padding: $std-dimension-gap;

  @media (min-width: std-get-breakpoint(480)) {
    width: 50%;
  }
}
