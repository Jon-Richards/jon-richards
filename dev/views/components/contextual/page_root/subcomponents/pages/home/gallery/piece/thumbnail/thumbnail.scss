/**
 * @fileoverview
 * Styling for the thumbnail component.
 */
 
@import '~scss-standards';

.root {
  display: flex;
  position: relative;
  width: 100%;
  height: 20rem;
  padding: 0rem;
  margin: 0rem;
  border: thin solid std-get-swatch(1);
  border-bottom-color: transparent;

  @media (min-width: std-get-breakpoint(720)) {
    max-width: 80%;
    margin: 0rem auto;
  }
}

.image {
  display: block;
  position: absolute;
  top: 0rem;
  left: 0rem;
  width: 100%;
  height: auto;
  z-index: 1;
  object-fit: cover;
}

.caption {
  display: flex;
  position: relative;
  z-index: 1;
  width: 100%;
  height: 100%;
  align-items: center;
  justify-content: center;
}
