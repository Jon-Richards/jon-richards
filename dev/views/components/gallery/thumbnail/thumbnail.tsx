/**
 * @fileoverview
 * Contains a thumbnail component for the gallery.
 */

import { React } from '../mediator';
const CSS = require('./thumbnail.scss');

/** Shape of a single thumbnail. */
export interface ThumbnailShape {
  /** This thumbnail's source image. */
  src: string;
  /** If the thumbnail should load its source. */
  shouldLoad: boolean;
  /** The thumbnail's width and height. */
  sizeModifier:
    | 'phone-vertical'
    | 'phone-horizontal'
    | 'tablet-vertical'
    | 'tablet-horizontal'
    | 'desktop'
    | 'banner';
  /** Alt text for the thumbnail's link. */
  altText: string;
  /** Method that fires when this thumbnail is clicked. */
  onClick?(): void;
}

/**
 *
 */
export const THUMBNAIL = React.memo<ThumbnailShape>(props => {
  const { src, shouldLoad, sizeModifier, altText, onClick = () => {} } = props;

  return (
    <figure
      role="button"
      aria-label="thumbnail"
      className={`${CSS['root']} ${CSS['root']}--${sizeModifier}`}
      onClick={() => onClick}
    >
      <img className={CSS['image']} src={src} alt={altText} />
    </figure>
  );
});
