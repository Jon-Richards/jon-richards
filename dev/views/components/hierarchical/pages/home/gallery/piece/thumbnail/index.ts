export { THUMBNAIL as Thumbnail } from './thumbnail';