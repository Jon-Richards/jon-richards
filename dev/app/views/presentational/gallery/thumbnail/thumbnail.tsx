/** @jsx jsx */

import * as React from 'react';
import { css, jsx } from '@emotion/core';
import { STYLES } from './styles';
import { Panel } from 'Views/presentational/panel';

/** Shape of a single thumbnail. */
export interface ThumbnailProps {
  /** This thumbnail's source image. */
  src: string;
  /** Alt text for the thumbnail's link. */
  altText: string;
  /** The destination for the thumbnail's link. */
  href: string;
  /** Method that fires when this thumbnail is clicked. */
  onClick(
    /** The HTML element's click event. */
    e: React.MouseEvent<HTMLElement, MouseEvent>
  ): void;
}

/** A single thumbnail. */
function Thumbnail (props: ThumbnailProps): JSX.Element {
  const { src, altText, href, onClick } = props;

  return (
    <a
      css={css(STYLES.root)}
      href={href}
      onClick={(e) => {
        e.preventDefault();
        onClick(e);
      }}
    >
      <Panel>
        <figure
          aria-label="thumbnail"
          css={css(STYLES.figure)}
        >
          <img css={css(STYLES.image)} src={src} alt={altText} />
          <figcaption css={css(STYLES.caption)}>{altText}</figcaption>
        </figure>
      </Panel>
    </a>
  );
}

const THUMBNAIL_MEMO = React.memo<ThumbnailProps>(Thumbnail);

export { THUMBNAIL_MEMO as Thumbnail };
