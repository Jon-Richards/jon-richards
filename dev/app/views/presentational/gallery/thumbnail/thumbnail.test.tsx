import * as React from 'react';
import { render, shallow } from 'enzyme';
import { Thumbnail } from './thumbnail';

describe('The gallery component.', () => {
  it('Should match the snapshot.', () => {
    const stubHandler = function () {};
    const test_1 = render(
      <Thumbnail
        src="test.jpg"
        altText="A test image."
        href="www.google.com"
        onClick={() => stubHandler()}
      />
    );

    expect(test_1).toMatchSnapshot();
  });

  it('Should fire the supplied click handler when clicked.', () => {
    const mockHandler = jest.fn();
    const test_1 = shallow(
      <Thumbnail
        src="test.jpg"
        altText="A test image."
        href="www.google.com"
        onClick={() => mockHandler()}
      />
    );

    test_1.find('[data-testid="gallery__thumb"]').simulate('click');
    expect(mockHandler).toBeCalledTimes(1);
  });
});
