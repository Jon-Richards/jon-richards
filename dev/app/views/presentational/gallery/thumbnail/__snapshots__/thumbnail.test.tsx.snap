// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The gallery component. Should match the snapshot. 1`] = `
.emotion-8 {
  display: inline-block;
  width: 100%;
  height: auto;
  padding: 0rem;
  margin: 0rem;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-7 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  width: 100%;
  height: auto;
  position: relative;
  padding: 1px;
  margin: 0rem;
  overflow: hidden;
  background-color: hsla(0,0%,14%,1);
  background-image: linear-gradient( 170deg, hsla(0,0%,30%,1), hsla(0,0%,10%,1), hsla(0,0%,17%,1) );
  box-shadow: 0.4rem 0.8rem 2rem 0rem hsla(0,0%,0%,0.5);
}

.emotion-6 {
  display: block;
  width: 100%;
  height: 100%;
  margin: 0rem;
  padding: 0rem;
  background-color: hsla(0,0%,14%,1);
  background-image: linear-gradient( 170deg, hsla(0,0%,14%,1), hsla(0,0%,8%,1) );
}

.emotion-5 {
  position: relative;
}

.emotion-1 {
  width: 100%;
  height: auto;
  padding: 0rem;
  margin: 0rem;
}

.emotion-0 {
  display: block;
  width: 100%;
  height: auto;
}

.emotion-4 {
  position: absolute;
  top: 0rem;
  left: 0rem;
  width: 100%;
  height: 100%;
  padding: 2rem;
  margin: 0rem;
  visibility: hidden;
}

.emotion-2 {
  color: hsla(48,7%,86%,1);
  margin-bottom: calc(2rem / 2);
  font-size: 1.2rem;
  font-family: "Open Sans",sans-serif;
  font-weight: 400;
}

.emotion-3 {
  font-size: 1rem;
  font-family: "Open Sans",sans-serif;
  font-weight: 300;
  line-height: 1.6rem;
  color: hsla(48,7%,86%,1);
}

<a
  class="emotion-8"
  data-testid="gallery__thumb"
  href="www.google.com"
>
  <div
    class="emotion-7"
  >
    <div
      class="emotion-6"
    >
      <div
        class="emotion-5"
      >
        <div
          class="emotion-1"
        >
          <picture>
            <source
              data-testid="picture__source"
              media="min-width(320px)"
              srcset="test_small.jpg"
            />
            <source
              data-testid="picture__source"
              media="min-width(480px)"
              srcset="test_medium.jpg"
            />
            <source
              data-testid="picture__source"
              media="min-width(720px)"
              srcset="test_large.jpg"
            />
            <img
              alt="Test 1 - A test image."
              class="emotion-0"
              src="test.jpg"
            />
          </picture>
        </div>
        <div
          class="emotion-4"
        >
          <h3
            class="emotion-2"
          >
            Test 1
          </h3>
          <p
            class="emotion-3"
          >
            A test image.
          </p>
        </div>
      </div>
    </div>
  </div>
</a>
`;
